src/main/java/io/pravega/connectors/flink/watermark/LowerBoundAssigner.java

```diff
@@ -0,0 +1,34 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.connectors.flink.watermark;
+
+import io.pravega.client.stream.TimeWindow;
+import org.apache.flink.streaming.api.watermark.Watermark;
+
+public abstract class LowerBoundAssigner<T> implements AssignerWithTimeWindows<T> {
+
+    private static final long serialVersionUID = 2069173720413829850L;
+
+    public LowerBoundAssigner() {
+    }
+
+    @Override
+    public abstract long extractTimestamp(T element, long previousElementTimestamp);
+
+    // built-in watermark implementation which emits the lower bound - 1
+    @Override
+    public Watermark getWatermark(TimeWindow timeWindow) {
+        if (timeWindow == null || timeWindow.isNearHeadOfStream()) {
```

Comment:
Could you add a little commentary in the source code about why `isNearHeadOfStream` is used here?  I don't understand it.

Reply:
```
    /**
     * Returns true if the reader is currently near the head of the stream and therefor no lower time bound can be obtained.
     */
    public boolean isNearHeadOfStream() {
        return lowerTimeBound == null;
    }
```
This is a check if the lowerbound is null, we don't emit a watermark.

Reply:
I've added a comment to LowerBoundAssigner